File: frontend/src/lib/api/ai.ts
import { api } from './client-safe';
import type { Trade } from './trades';

// AI Analysis Types
export interface TradeScore {
  overall_score: number; // 0-100
  execution_score: number;
  timing_score: number;
  strategy_score: number;
  risk_reward_ratio: number;
  pnl: number;
  pnl_percentage: number;
  insights: string[];
  recommendations: string[];
  market_context: MarketContext;
}

export interface TradeCritique {
  summary: string;
  suggestion: string;
  confidence: number; // 1-10
  tags: string[];
  technical_analysis: string;
  psychological_analysis: string;
  risk_assessment: string;
}

export interface BehavioralInsights {
  emotional_state: string;
  consistency_rating: number;
  discipline_score: number;
  risk_profile: string;
  patterns_detected: PatternDetection[];
  improvement_areas: string[];
  streaks: StreakAnalysis;
}

export interface PatternDetection {
  pattern_type: string;
  frequency: number;
  impact_on_pnl: number;
  description: string;
  examples: string[];
  recommendations: string;
}

export interface StreakAnalysis {
  current_streak: number;
  streak_type: 'winning' | 'losing' | 'neutral';
  best_streak: number;
  worst_streak: number;
  average_streak_length: number;
}

export interface EdgeStrength {
  strategy: string;
  win_rate: number;
  profit_factor: number;
  sharpe_ratio: number;
  kelly_criterion: number;
  sample_size: number;
  confidence_level: number;
  market_conditions: string[];
}

export interface MarketContext {
  regime: 'bull' | 'bear' | 'sideways';
  volatility: 'low' | 'medium' | 'high';
  trend_strength: number;
  support_levels: number[];
  resistance_levels: number[];
  recommendation: string;
}

export interface EmotionalAnalytics {
  dominant_emotions: Array<{
    emotion: string;
    frequency: number;
    impact: number;
  }>;
  emotional_consistency: number;
  best_performing_emotion: string;
  worst_performing_emotion: string;
  recommendations: string[];
}

export interface PreTradeAnalysis {
  should_take_trade: boolean;
  confidence_score: number;
  risk_score: number;
  pattern_matches: Array<{
    pattern: string;
    similarity: number;
    historical_outcome: string;
  }>;
  market_alignment: boolean;
  psychological_readiness: number;
  suggested_position_size: number;
  warnings: string[];
}

export interface AIInsightsSummary {
  trade_score: TradeScore;
  critique: TradeCritique;
  behavioral_insights: BehavioralInsights;
  edge_strength: EdgeStrength[];
  market_context: MarketContext;
  emotional_analytics: EmotionalAnalytics;
}

// AI API Service
export const aiApi = {
  // Trade Intelligence
  async getTradeScore(tradeId: string): Promise<TradeScore> {
    return api.get(`/api/v1/ai/trades/${tradeId}/score`);
  },

  async getTradeScoreBulk(tradeIds: string[]): Promise<Record<string, TradeScore>> {
    return api.post('/api/v1/ai/trades/score/bulk', { trade_ids: tradeIds });
  },

  // Trade Critique
  async getTradeCritique(tradeId: string): Promise<TradeCritique> {
    return api.get(`/api/v1/critique/trades/${tradeId}`);
  },

  async regenerateCritique(tradeId: string): Promise<TradeCritique> {
    return api.post(`/api/v1/critique/trades/${tradeId}/regenerate`);
  },

  // Behavioral Analytics
  async getBehavioralInsights(timeframe: 'week' | 'month' | 'quarter' | 'year' = 'month'): Promise<BehavioralInsights> {
    return api.get(`/api/v1/ai/behavioral/insights?timeframe=${timeframe}`);
  },

  async getPatternDetection(): Promise<PatternDetection[]> {
    return api.get('/api/v1/ai/patterns/detect');
  },

  // Edge Strength Analysis
  async getEdgeStrength(): Promise<EdgeStrength[]> {
    return api.get('/api/v1/ai/edge/strength');
  },

  async getEdgeByStrategy(strategy: string): Promise<EdgeStrength> {
    return api.get(`/api/v1/ai/edge/strategy/${strategy}`);
  },

  // Market Context
  async getMarketContext(symbol?: string): Promise<MarketContext> {
    return api.get('/api/v1/ai/market/context', { params: { symbol } });
  },

  async getMarketRegime(): Promise<{ regime: string; confidence: number; factors: string[] }> {
    return api.get('/api/v1/ai/market/regime');
  },

  // Emotional Analytics
  async getEmotionalAnalytics(timeframe: 'week' | 'month' | 'quarter' = 'month'): Promise<EmotionalAnalytics> {
    return api.get(`/api/v1/ai/emotional/analytics?timeframe=${timeframe}`);
  },

  async getEmotionalImpact(): Promise<any> {
    return api.get('/api/v1/ai/emotional/impact');
  },

  // Pre-Trade Analysis
  async analyzePreTrade(trade: Partial<Trade>): Promise<PreTradeAnalysis> {
    return api.post('/api/v1/ai/pre-trade/analyze', trade);
  },

  // AI Insights Summary
  async getAIInsightsSummary(): Promise<AIInsightsSummary> {
    return api.get('/api/v1/ai/insights/summary');
  },

  // Pattern Recognition
  async detectPatterns(timeframe: 'week' | 'month' | 'all' = 'month'): Promise<PatternDetection[]> {
    return api.get(`/api/v1/ai/patterns?timeframe=${timeframe}`);
  },

  async getPatternHistory(patternType: string): Promise<any> {
    return api.get(`/api/v1/ai/patterns/${patternType}/history`);
  },

  // AI Recommendations
  async getRecommendations(): Promise<{
    immediate: string[];
    weekly: string[];
    improvement_plan: string[];
  }> {
    return api.get('/api/v1/ai/recommendations');
  },

  // Risk Analysis
  async getRiskAnalysis(): Promise<{
    current_risk_score: number;
    var_95: number;
    var_99: number;
    max_drawdown_predicted: number;
    position_sizing_recommendation: number;
    warnings: string[];
  }> {
    return api.get('/api/v1/ai/risk/analysis');
  },

  // Real-time scoring via WebSocket
  subscribeToLiveScoring(onScore: (score: TradeScore) => void): () => void {
    // This would connect to WebSocket for real-time updates
    // Implementation depends on WebSocket setup
    return () => {
      // Cleanup function
    };
  }
};

File: services/gateway/backend/schemas/ai.py
"""
AI Intelligence Schemas
"""
from typing import List, Dict, Any, Optional
from datetime import datetime, date
from pydantic import BaseModel, Field


class TradeScoreResponse(BaseModel):
    overall_score: float = Field(..., ge=0, le=100)
    execution_score: float = Field(..., ge=0, le=100)
    timing_score: float = Field(..., ge=0, le=100)
    strategy_score: float = Field(..., ge=0, le=100)
    risk_reward_ratio: float
    pnl: float
    pnl_percentage: float
    insights: List[str]
    recommendations: List[str]
    market_context: Dict[str, Any]


class TradeCritiqueResponse(BaseModel):
    summary: str
    suggestion: str
    confidence: int = Field(..., ge=1, le=10)
    tags: List[str]
    technical_analysis: str
    psychological_analysis: str
    risk_assessment: str


class PatternDetection(BaseModel):
    pattern_type: str
    frequency: int
    impact_on_pnl: float
    description: str
    examples: List[str]
    recommendations: str


class StreakAnalysis(BaseModel):
    current_streak: int
    streak_type: str = Field(..., pattern="^(winning|losing|neutral)$")
    best_streak: int
    worst_streak: int
    average_streak_length: float


class BehavioralInsightsResponse(BaseModel):
    emotional_state: str
    consistency_rating: float
    discipline_score: float
    risk_profile: str
    patterns_detected: List[PatternDetection]
    improvement_areas: List[str]
    streaks: StreakAnalysis


class EdgeStrengthResponse(BaseModel):
    strategy: str
    win_rate: float
    profit_factor: float
    sharpe_ratio: float
    kelly_criterion: float
    sample_size: int
    confidence_level: float
    market_conditions: List[str]


class PatternDetectionResponse(BaseModel):
    pattern_type: str
    frequency: int
    impact_on_pnl: float
    description: str
    examples: List[str]
    recommendations: str


class MarketContextResponse(BaseModel):
    regime: str = Field(..., pattern="^(bull|bear|sideways)$")
    volatility: str = Field(..., pattern="^(low|medium|high)$")
    trend_strength: float
    support_levels: List[float]
    resistance_levels: List[float]
    recommendation: str


class EmotionMetric(BaseModel):
    emotion: str
    frequency: float
    impact: float


class EmotionalAnalyticsResponse(BaseModel):
    dominant_emotions: List[EmotionMetric]
    emotional_consistency: float
    best_performing_emotion: str
    worst_performing_emotion: str
    recommendations: List[str]


class PatternMatch(BaseModel):
    pattern: str
    similarity: float
    historical_outcome: str


class PreTradeAnalysisRequest(BaseModel):
    symbol: str
    side: str = Field(..., pattern="^(long|short)$")
    entry_price: float
    quantity: int
    stop_loss: Optional[float] = None
    take_profit: Optional[float] = None
    strategy: Optional[str] = None
    confidence_level: Optional[int] = Field(None, ge=1, le=10)


class PreTradeAnalysisResponse(BaseModel):
    should_take_trade: bool
    confidence_score: float
    risk_score: float
    pattern_matches: List[PatternMatch]
    market_alignment: bool
    psychological_readiness: float
    suggested_position_size: float
    warnings: List[str]


class AIInsightsSummaryResponse(BaseModel):
    trade_score: TradeScoreResponse
    critique: TradeCritiqueResponse
    behavioral_insights: BehavioralInsightsResponse
    edge_strength: List[EdgeStrengthResponse]
    market_context: MarketContextResponse
    emotional_analytics: EmotionalAnalyticsResponse


class RecommendationsResponse(BaseModel):
    immediate: List[str]
    weekly: List[str]
    improvement_plan: List[str]


class RiskAnalysisResponse(BaseModel):
    current_risk_score: float
    var_95: float
    var_99: float
    max_drawdown_predicted: float
    position_sizing_recommendation: float
    warnings: List[str]